pymc3/step_methods/smc_utils.py

```diff
@@ -78,3 +85,197 @@ def _posterior_to_trace(posterior, variables, model, var_info):
+    weights : numpy array
+        Importance weights (floats)
+    sj : float
+        Partial marginal likelihood
+    """
+    low_beta = old_beta = beta
+    up_beta = 2.0
+    rN = int(len(likelihoods) * threshold)
+
+    while up_beta - low_beta > 1e-6:
+        new_beta = (low_beta + up_beta) / 2.0
+        weights_un = np.exp((new_beta - old_beta) * (likelihoods - likelihoods.max()))
+        weights = weights_un / np.sum(weights_un)
+        ESS = int(1 / np.sum(weights ** 2))
+        if ESS == rN:
+            break
+        elif ESS < rN:
+            up_beta = new_beta
+        else:
+            low_beta = new_beta
+    if new_beta >= 1:
+        new_beta = 1
+    sj = np.exp((new_beta - old_beta) * likelihoods)
+    weights_un = np.exp((new_beta - old_beta) * (likelihoods - likelihoods.max()))
+    weights = weights_un / np.sum(weights_un)
+    return new_beta, old_beta, weights, np.mean(sj)
+
+
+def logp_forw(out_vars, vars, shared):
+    """Compile Theano function of the model and the input and output variables.
+
+    Parameters
+    ----------
+    out_vars : List
+        containing :class:`pymc3.Distribution` for the output variables
+    vars : List
+        containing :class:`pymc3.Distribution` for the input variables
+    shared : List
+        containing :class:`theano.tensor.Tensor` for depended shared data
+    """
+    out_list, inarray0 = join_nonshared_inputs(out_vars, vars, shared)
+    f = theano.function([inarray0], out_list[0])
+    f.trust_input = True
+    return f
+
+
+class PseudoLikelihood:
+    """
+    Pseudo Likelihood
+    """
+
+    def __init__(
+        self,
+        epsilon,
+        observations,
+        function,
+        model,
+        var_info,
+        distance="absolute_error",
+        sum_stat=False,
+    ):
+        """
+        kernel : function
+            a valid scipy.stats distribution. Defaults to `stats.norm`
+
+        """
+        self.epsilon = epsilon
+        self.observations = observations
+        self.function = function
+        self.model = model
+        self.var_info = var_info
+        self.kernel = self.gauss_kernel
+        self.dist_func = distance
+        self.sum_stat = sum_stat
+
+        if distance == "absolute_error":
+            self.dist_func = self.absolute_error
+        elif distance == "sum_of_squared_distance":
+            self.dist_func = self.sum_of_squared_distance
+        else:
+            raise ValueError("Distance metric not understood")
+
+    def posterior_to_function(self, posterior):
+        model = self.model
+        var_info = self.var_info
+        parameters = {}
+        size = 0
+        for var, values in var_info.items():
+            shape, new_size = values
+            value = posterior[size : size + new_size].reshape(shape)
+            if is_transformed_name(var):
+                var = get_untransformed_name(var)
+                value = model[var].transformation.backward_val(value)
```

Comment:
can't you do `model[var].transformation.backward(value).eval()`?

Reply:
Yes, but this is faster.

Reply:
Couldn't you compile and store a `theano.function` of `model[var].transformation.backward` to make it even faster? I'd prefer to avoid the redundancy of the transforms.

Reply:
I think this is not redundant, we already have `forward` and `forward_val`, so it seems reasonable to have `backward` and `backward_val`.

Reply:
Alright then.